File: sdk/go/protomcp/completion.go
package protomcp

// CompletionResult holds the result of a completion handler.
type CompletionResult struct {
	Values  []string
	Total   int32
	HasMore bool
}

// completionKey uniquely identifies a completion handler.
type completionKey struct {
	RefType string
	RefName string
	ArgName string
}

// CompletionHandler takes the current argument value and returns completions.
type CompletionHandler func(argumentValue string) CompletionResult

var completionRegistry = map[completionKey]CompletionHandler{}

// RegisterCompletion registers a completion handler keyed by (refType, refName, argName).
func RegisterCompletion(refType, refName, argName string, handler CompletionHandler) {
	completionRegistry[completionKey{RefType: refType, RefName: refName, ArgName: argName}] = handler
}

// GetCompletionHandler looks up a completion handler by key.
func GetCompletionHandler(refType, refName, argName string) (CompletionHandler, bool) {
	h, ok := completionRegistry[completionKey{RefType: refType, RefName: refName, ArgName: argName}]
	return h, ok
}

// ClearCompletionRegistry resets the completion registry.
func ClearCompletionRegistry() {
	completionRegistry = map[completionKey]CompletionHandler{}
}

File: sdk/python/src/protomcp/completion.py
from dataclasses import dataclass
from typing import Callable

_completion_registry: dict[tuple[str, str, str], Callable] = {}

@dataclass
class CompletionResult:
    values: list[str]
    total: int = 0
    has_more: bool = False

def completion(ref_type: str, ref_name: str, argument_name: str):
    """Register a completion provider for a specific ref+argument combination.

    ref_type: "ref/prompt" or "ref/resource"
    ref_name: name of the prompt or URI of the resource
    argument_name: which argument to complete
    """
    def decorator(func: Callable) -> Callable:
        _completion_registry[(ref_type, ref_name, argument_name)] = func
        return func
    return decorator

def get_completion_handler(ref_type: str, ref_name: str, argument_name: str):
    return _completion_registry.get((ref_type, ref_name, argument_name))
